### PR TITLE
fix textWidth() and textToPoints()

### LINKED ISSUE #5036: textWidth() returns wrong value for strings with "\n"
#### Most appropriate sub-area of p5.js?

- [ ] Accessibility (Web Accessibility)
- [ ] Build tools and processes
- [ ] Color
- [ ] Core/Environment/Rendering
- [x] Data
- [ ] DOM
- [ ] Events
- [ ] Friendly error system
- [ ] Image
- [ ] IO (Input/Output)
- [ ] Localization
- [ ] Math
- [ ] Unit Testing
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Other (specify if possible)

#### Details about the bug:

- p5.js version: p5@1.2.0
- Web browser and version: Google Chrome | 88.0.4324.146
- Operating System: Browser on Windows 10, web server on Raspbian
- Steps to reproduce this:
```
    let foo = "hello";
    let bar = "world";
    print("foo: " + pg.textWidth(foo));
    print("bar: " + pg.textWidth(bar));

    //bug is here
    let foobar = foo + "\n" + bar;
    print("foobar: " + pg.textWidth(foobar));

    foobar = foo + " " + bar;
    print("foobar: " + pg.textWidth(foobar));
```

here the textWidth() of foobar should be the same size as the widest text foo or bar, in this case, bar
Instead, the current behavior is ignoring the interpreting the "\n" as a " "(space)

Running the code above I'm getting this:
```
foo: 31.69189453125
bar: 35.8447265625
foobar: 71.7041015625
foobar: 71.7041015625
```

### LINKED ISSUE #5474: textToPoints() not handling line-breaks correctly
textToPoints() inserts a blank character instead of a line-break

For the string 'B\nA', text() (in white) handles linebreak correctly, while textToPoints() does not:

image

```js

// load font first

function setup() {
  createCanvas(400, 400);
  textFont(font, 100);
  pts = font.textToPoints('B\nA', 0, 0, 100, {
    sampleFactor: 0.4 // increase for more points
  });
  background(0);
  fill(255);
  noLoop()
}

function draw() {
  text('B\nA', 200, 200)
  pts.forEach(p => circle(200 + p.x, 200 + p.y, 3));
}
```

### PR DESCRIPTION
**Pull request by munusshih**
Resolves #5036 
Resolves #5474 

 Changes:

- Both `textWidth()` and `textToPoints()` have trouble dealing with `\n` or `\t`, [escape character](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Character_escape)
- Or when there’s line break in [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)
- I fixed both of them without changing `opentype.js`

- Try the new `textWidth` [here](https://editor.p5js.org/munusshih/sketches/yDdgjeBXl)
- Try the new `textToPoints` [here](https://editor.p5js.org/munusshih/sketches/xNUOwmP1h)

 Screenshots of the change:

- Black box: the original `textWidth()`
- Red box: the fixed `textWidth()`
Screenshot 2023-06-06 at 7 09 33 PM

- Original `textToPoints()` breaks down when encounter line breaks or tabs
Screenshot 2023-06-06 at 7 14 34 PM

- Fixed `textToPoints()` handle them correctly
Screenshot 2023-06-06 at 7 18 13 PM

#### PR Checklist

- [x] `npm run lint` passes
- [x] [Inline documentation] is included / updated

[Inline documentation]: https://github.com/processing/p5.js/blob/main/contributor_docs/inline_documentation.md
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

### CODE REVIEW

**Comment by Qianqianye** [comment]
Thanks @munusshih. I am inviting the GSoC mentors @aahdee and @kyeah, as well as the advisor @limzykenneth to review this PR.

File: src/typography/p5.Font.js
Comment (limzykenneth) [inline src/typography/p5.Font.js:479]: What is the rendering issue you are mitigating here?

**Comment by limzykenneth** [comment]
@munusshih Not sure how practical it would be but if you can add some unit tests that could work cross platform it would help prevent future regression. Although it might be difficult if we consider possible different system this may run on, so if not that's fine as well.

@aahdee @kyeah Do have a look at this and if you are happy with it as well we can merge after resolving the line comment on the latest commit.

File: src/typography/p5.Font.js
Comment (kyeah) [reply src/typography/p5.Font.js:479]: Correct me if I'm wrong @munusshih but I believe this is just avoiding the textWidth calculation if we don't need the `textWidth` value 🙂 The variable is only used below if we have CENTER or RIGHT alignment. There's no functional change here.

**Review by kyeah — COMMENTED** [review]
Looks good overall! I made some minor stylistic comments but mainly interested in resolving the issue with different character widths for `textWidth()`.

Agreed with @limzykenneth that adding some unit tests would be beneficial. I see there's very minimal unit test coverage in the `tests/typography/p5.Font.js` file currently though 😅 so I'd be okay merging this in without unit tests.

If you want to try adding a unit test or a couple of unit tests, there's one `textBounds` example that I think can provide some inspiration for textWidth unit tests. It loads a custom test font so I think the test would behave the same across platforms. If you're interested but need some help we can try to look at it together next week.

**Comment by Qianqianye** [comment]
Thanks @limzykenneth  and @kyeah for reviewing. @munusshih, would it be possible to make some updates based on the comments? Thanks!

**Comment by munusshih** [comment]
Yes! I've updated the code and is waiting for approval now :)

**Comment by limzykenneth** [comment]
@kyeah @aahdee Do have a look and if you are both happy with this I will merge. Thanks.

File: src/typography/attributes.js
Comment (kyeah) [inline src/typography/attributes.js:242]:
```suggestion
  // Return the textWidth for every line
```

typo 👋

File: src/typography/attributes.js
Comment (kyeah) [inline src/typography/attributes.js:224]: I think changing the line length from 50px to 100px would still improve the documentation @munusshih!

**Review by kyeah — APPROVED** [review]
Thanks @munusshih! works well now 😄 

You could probably avoid creating an array if you wanted to, but I think it's probably not a noticeable difference in performance unless people pass in an entire novel's worth of text 😂 Thinking of something like this to replace `newArr`:

```js
const largestWidth = 0;

for(let i=0; i<textLines.length; i++){
  // calculate line width and update largestWidth if needed
  ...
}

return largestWidth;
```

**Comment by limzykenneth** [comment]
Looks good. Thanks!

**Pull request merged**
